### PR TITLE
[FLINK-17568][checkpointing] Fix the issue that task may consume input data after checkpoint barrier before checkpoint is performed for unaligned checkpoint

### PR DESCRIPTION
**Pull request by wsry**
## What is the purpose of the change
For unaligned checkpoint, task may consume data after the checkpoint barrier before performing checkpoint which lead to consumption of duplicated data and corruption of data stream.

More specifically, when the Netty thread notifies the checkpoint barrier for the first time and enqueue a checkpointing task in the mailbox, the task thread may still in data consumption loop and if it reads a new checkpoint barrier from another channel it will not return to the mailbox and instead it will continue to read data until a all data consumed or we have a full record, meanwhile, the data after checkpoint barrier may be read and consumed which lead to inconsistency.

This PR tries to fix the problem.

## Brief change log

  - Return to mailbox after receiving a checkpoint barrier.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wsry** [comment]
@pnowojski I have added a unit test to ```StreamTaskNetworkInputTest```, could you please take a look and see whether I understand correctly and the test case verifies what you want to verify?

**Comment by wsry** [comment]
@pnowojski 
I found that the changed behavior of CheckpointedInputGate fails CheckpointBarrierTrackerTest and CreditBasedCheckpointBarrierAlignerTest. I have fixed them accordingly. 
Besides, I found that CheckpointBarrierAlignerTestBase#testMultiChannelSkippingCheckpoints is a duplication of CheckpointBarrierAlignerTestBase#testMultiChannelJumpingOverCheckpoint, so I removed CheckpointBarrierAlignerTestBase#testMultiChannelSkippingCheckpoints in a hotfix commit.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInputTest.java [hunk 1]
```diff
@@ -74,29 +77,31 @@ public void tearDown() throws Exception {
...
-			new StatusWatermarkValve(1, output),
-			0);
+		StreamTaskNetworkInput input = createStreamTaskNetworkInput(buffers, output);
 
 		assertHasNextElement(input, output);
 		assertHasNextElement(input, output);
 		assertEquals(2, output.getNumberOfEmittedRecords());
 	}
 
+	@Test
```
Comment (pnowojski) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/StreamTaskNetworkInputTest.java:90]: nit: add an explanation comment:
```
InputGate on CheckpointBarrier can enqueue a mailbox action to execute and StreamTaskNetworkInput must
allow this action to execute before processing a following record.
```

**Comment by wsry** [comment]
@pnowojski I have updated the PR according to the review comments.

**Pull request merged**